File: BBMTLib/tss/descriptor_checksum.go
package tss

import (
	"fmt"
	"strings"
)

// Output descriptor checksum (BIP 380 / Bitcoin Core descriptor.cpp).

const descriptorInputCharset = "0123456789()[],'/*abcdefgh@:$%{}" +
	"IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~" +
	"ijklmnopqrstuvwxyzABCDEFGH`#\"\\ "
const descriptorChecksumCharset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"

var descriptorGenerator = [5]uint64{0xf5dee51989, 0xa9fdca3312, 0x1bab10e32d, 0x3706b1677a, 0x644d626ffd}

func descriptorPolyMod(c uint64, val int) uint64 {
	c0 := byte(c >> 35)
	c = ((c & 0x7ffffffff) << 5) ^ uint64(val)
	if c0&1 != 0 {
		c ^= descriptorGenerator[0]
	}
	if c0&2 != 0 {
		c ^= descriptorGenerator[1]
	}
	if c0&4 != 0 {
		c ^= descriptorGenerator[2]
	}
	if c0&8 != 0 {
		c ^= descriptorGenerator[3]
	}
	if c0&16 != 0 {
		c ^= descriptorGenerator[4]
	}
	return c
}

func descriptorExpandSymbols(desc string) ([]int, error) {
	symbols := make([]int, 0, len(desc)+3)
	cls := 0
	clsCount := 0
	for _, ch := range desc {
		pos := strings.IndexRune(descriptorInputCharset, ch)
		if pos < 0 {
			return nil, fmt.Errorf("invalid descriptor character %q", ch)
		}
		symbols = append(symbols, pos&31)
		cls = cls*3 + (pos >> 5)
		clsCount++
		if clsCount == 3 {
			symbols = append(symbols, cls)
			cls = 0
			clsCount = 0
		}
	}
	if clsCount > 0 {
		symbols = append(symbols, cls)
	}
	return symbols, nil
}

// DescriptorChecksum returns the 8-character checksum for a descriptor body (no # suffix).
func DescriptorChecksum(desc string) (string, error) {
	symbols, err := descriptorExpandSymbols(desc)
	if err != nil {
		return "", err
	}
	c := uint64(1)
	for _, sym := range symbols {
		c = descriptorPolyMod(c, sym)
	}
	for i := 0; i < 8; i++ {
		c = descriptorPolyMod(c, 0)
	}
	c ^= 1
	out := make([]byte, 8)
	for j := 0; j < 8; j++ {
		out[j] = descriptorChecksumCharset[(c>>(5*(7-j)))&31]
	}
	return string(out), nil
}

// AppendOutputDescriptorChecksum appends Sparrow-style #checksum (gomobile / Android Tss API).
func AppendOutputDescriptorChecksum(desc string) (string, error) {
	return AddDescriptorChecksum(desc)
}

// AddDescriptorChecksum appends #CHECKSUM to a descriptor body.
func AddDescriptorChecksum(desc string) (string, error) {
	if desc == "" {
		return "", nil
	}
	body := desc
	if i := strings.LastIndex(desc, "#"); i >= 0 {
		body = desc[:i]
	}
	checksum, err := DescriptorChecksum(body)
	if err != nil {
		return "", err
	}
	return body + "#" + checksum, nil
}

File: android/app/src/main/java/com/boldwallet/DescriptorChecksum.kt
package com.boldwallet

/**
 * Sparrow / Bitcoin Core output-descriptor checksum (BIP 380).
 * Mirrors [github.com/BoldBitcoinWallet/BBMTLib/tss/descriptor_checksum.go].
 */
object DescriptorChecksum {
    private const val INPUT_CHARSET =
        "0123456789()[],'/*abcdefgh@:$%{}" +
            "IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~" +
            "ijklmnopqrstuvwxyzABCDEFGH`#\"\\ "
    private const val CHECKSUM_CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"

    private val GENERATOR = longArrayOf(
        0xf5dee51989L,
        0xa9fdca3312L,
        0x1bab10e32dL,
        0x3706b1677aL,
        0x644d626ffdL,
    )

    fun appendToDescriptor(desc: String): String {
        if (desc.isEmpty()) return desc
        val body = if (desc.contains('#')) desc.substringBeforeLast('#') else desc
        val checksum = checksum(body)
        return "$body#$checksum"
    }

    private fun checksum(desc: String): String {
        val symbols = expandSymbols(desc)
        var c = 1L
        for (sym in symbols) {
            c = polyMod(c, sym)
        }
        repeat(8) { c = polyMod(c, 0) }
        c = c xor 1L
        return buildString(8) {
            for (j in 0 until 8) {
                val idx = ((c shr (5 * (7 - j))) and 31L).toInt()
                append(CHECKSUM_CHARSET[idx])
            }
        }
    }

    private fun polyMod(c: Long, value: Int): Long {
        val c0 = (c shr 35).toByte()
        var out = ((c and 0x7ffffffffL) shl 5) xor value.toLong()
        if (c0.toInt() and 1 != 0) out = out xor GENERATOR[0]
        if (c0.toInt() and 2 != 0) out = out xor GENERATOR[1]
        if (c0.toInt() and 4 != 0) out = out xor GENERATOR[2]
        if (c0.toInt() and 8 != 0) out = out xor GENERATOR[3]
        if (c0.toInt() and 16 != 0) out = out xor GENERATOR[4]
        return out
    }

    private fun expandSymbols(desc: String): IntArray {
        val symbols = mutableListOf<Int>()
        var cls = 0
        var clsCount = 0
        for (ch in desc) {
            val pos = INPUT_CHARSET.indexOf(ch)
            if (pos < 0) {
                throw IllegalArgumentException("invalid descriptor character '$ch'")
            }
            symbols.add(pos and 31)
            cls = cls * 3 + (pos shr 5)
            clsCount++
            if (clsCount == 3) {
                symbols.add(cls)
                cls = 0
                clsCount = 0
            }
        }
        if (clsCount > 0) {
            symbols.add(cls)
        }
        return symbols.toIntArray()
    }
}
